### DISCUSSION TITLE
Feature request: Select All

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/OdfEdit#25] [Ideas]
Please add a capability to select the whole text window content with pressing `Ctrl+A` as like other applications do.

Also if some text is already selected, paste from clipboard should replace the selected text with the clipboard content. But now the text from clipboaard is added instead of replacing. It would be better to fix it.

**Comment by eturpault** [comment]
What you are requesting is already working like this in Windows (I use the Tkinter GUI framework). It seems that Tkinter is not behaving in the same way in Linux. According to what I see in some forums it has to be implemented manually. I will do it.

**Comment by eturpault** [comment]
It is implemented for OdfEdit v2.5.
Here is a beta release with this fix (and the drag&drop feature as well).
[OdfEdit-2.5beta1.zip](https://github.com/GrandOrgue/ODFEdit/files/11778979/OdfEdit-2.5beta1.zip)

**Comment by oleg68** [comment]
@eturpault I'm starting to test OdfEdit-2.5 beta now.

I'll put my notes here.

**Comment by oleg68** [comment]
1. The drag-and-drop capability is great, thank you. But I'm able to select only one rank i a time. I'd like to have a capability to select several ranks in a time and move them to another windchest at once.

**Reply by eturpault** [reply]
Yes it would be good to have the possiblity to move several objects at the same time, I tried to implement it but I encountered issues. In the objects tree if one selects several items, the childs of the selected objects are selected as well, even if they are not visible, which is not good. Probably it should be possible to select and drag multiple objects from the objects list only, I will try to make it possible (only if all the selected objects are of the same type). Else objects moving is not an operation done often and there are never a lot of objects to move, moving one by one takes a reasonable time.

**Comment by oleg68** [comment]
2. When I add a windchest with the `Add` button, an invalid empty windchest appears:

```
[WindchestGroup013]
Name=WindchestGroup template
NumberOfEnclosures=0
Enclosure001=
NumberOfTremulants=0
Tremulant001=
```

I'd not put `Enclosure001=` and `Tremulant001=` here.

**Reply by eturpault** [reply]
Yes there are certainly improvements to do in my templates, I will integrate this one. They can be modified in the file GoObjectsTemplates.txt located in the folder resources. If you see other improvements to do, with the experience that you have, feel free to modify this file and share it to me, thank you.

**Comment by oleg68** [comment]
3. When converting a Erfurt Predigerkirche (demo) from hauptwerk to odf, I receive

```
[Rank034]
Name=105. P  Pommer 4' (front) UNUSED
...
Pipe029PitchTuning=203
...
Pipe031PitchTuning=403
...
Pipe032PitchTuning=203
```

I doubt that 203, 403, 203 are correct. I expect 200, 400, 200.

**Reply by eturpault** [reply]
For example for the Pipe029 (MIDI note 64), in the HW ODF there is : 
- at pipe level : Pitch_OriginalOrgan_PitchHz=651.76357844136
- at sample level : Pitch_ExactSamplePitch=579.59

Between 579.59Hz and 651.76357844136Hz there are 203 cents. I could round the result of the calculation, but is it worth to do it ? Better to be precise when one have the exact data.

**Reply by eturpault** [reply]
The formula that I use to convert a frequency to cents is : 
`cents = 1200 * log2( frequency / reference_frequency )`
reference_frequency is 440Hz by default, else I use the base pitch frequency of the organ if it is provided in the HW ODF. It is 435Hz in the Erfurt sample set.

**Reply by oleg68** [reply]
Because Pipe029 borrows Pipe027 samples and you don't write Pipe027PitchTuning, its value should also be taken into account. The right formula is:

Pipe029PitchTuning = 1200 * log2( Pipe029frequency / Pipe029reference_frequency * Pipe027reference_frequency/Pipe027frequency )

**Reply by eturpault** [reply]
This is certainly the right formula in another use case, but in my case I cannot use it. I convert HW pipes to GO pipes one by one and independently from the other pipes of the same rank. The pitch tuning can be done only between the frequency of the current converted pipe and the frequency of its attached sample (which can be indeed borrowed to another pipe). 
Playing MIDI notes 62-63-64 of this Pommer 4' rank with GO gives a good audio result I think, the MIDI note 64 doesn't feel to be detuned compared to the others.

**Reply by oleg68** [reply]
But your formula is not correct for borrowed pipes

**Reply by oleg68** [reply]
> I convert HW pipes to GO pipes one by one and independently from the other pipes of the same rank

I suggest to cache pipe information for using it with another pipes.

**Reply by eturpault** [reply]
I would like to apply the formula that you are recommanding, but sorry I still don't see how to do it.
Here are the pitch data that I recover from the HW sample set, still with the same example of the Pedal Pommer 4' rank, MIDI notes 62 and 64.

MIDI 62 / Pipe027 : 

- OriginalOrgan_PitchHz=579.59 
- linked to the sample "P Pommer 4/A0_D/062-d.wav" with ExactSamplePitch=579.59

MIDI 64 / Pipe029 : 

- OriginalOrgan_PitchHz=651.76357844136 
- linked to the sample of Pipe027

Organ base pitch frequency=435Hz

Which should be Pipe029frequency, Pipe029reference_frequency, Pipe027reference_frequency and Pipe027frequency values according to you ?

Is it correct to add the following parenthesis around each pipe frequencies division ?
Pipe029PitchTuning = 1200 * log2( (Pipe029frequency / Pipe029reference_frequency) * (Pipe027reference_frequency / Pipe027frequency) )

**Reply by oleg68** [reply]
I took your formula

> cents = 1200 * log2( frequency / reference_frequency )

My formula is to calculating `cents` for Pipe027 and Pipe029 separratelly and then to substract them. Because log2(X)-log2(Y) = log2(X/Y), I wrote my formula.

> Which should be Pipe029frequency, Pipe029reference_frequency, Pipe027reference_frequency and Pipe027frequency

They are the same as you wrote above, but for different pipes. I don't know where are you get these values from in xml.

> Is it correct to add the following parenthesis around each pipe frequencies division ?
Pipe029PitchTuning = 1200 * log2( (Pipe029frequency / Pipe029reference_frequency) * (Pipe027reference_frequency / Pipe027frequency) )

Yes, of cause.

**Reply by eturpault** [reply]
Ok clear, thanks. So I am already calculating the pitch tuning as you recommand (i.e. comparing the cents of the two pipes) but doing it in a different way : log2(X)-log2(Y) instead of log2(X/Y) [ not log2(X*Y) ].
I have changed my code to use your formula, it is more optimized. 

Pipe029PitchTuning = 1200 * log2( (651.76357844136 / 435) / (579.59 / 435) ) = 203

**Reply by oleg68** [reply]
Dividing by 435 is not necessary: (a/c)/(b/c) = a/b

**Reply by eturpault** [reply]
You are right, even simpler formula.

**Reply by oleg68** [reply]
But 435 should be converted to the organ level PitchCorrection value: 1200 * log2(435/440)

**Reply by eturpault** [reply]
Done. Now OdfEdit will make right pitch conversion at organ and pipes levels. Thanks.

**Comment by oleg68** [comment]
4. There is a ``Search in ODF`` window. But sometimes it is necessary to search in the current section only. By example, search all 'PipeXXXPitchTuning` occurences.

**Reply by eturpault** [reply]
In the search results list, at the beginning of each line there is the section name, and the results are sorted by section name, so you can easily see together in the list all the occurences of the searched text of the same section. Is it what you need ? Else I have misunderstood your comment.

**Comment by oleg68** [comment]
5. A capability of Find and Replace is higly desired (for the current section, for all subobjects of the current or for the all ODF). For example, replace all occurences `..\OrganInstallationPackages\002524\` to `pipes/erfurt/`.

**Reply by eturpault** [reply]
Good idea, I will add it.

**Reply by kerkovits** [reply]
Regex possibility for search&reaplace would definitely be a good option for more complex use cases, if it is not a large overhead.

**Reply by eturpault** [reply]
Python has a dedicated module to manage Regex expressions, so it should not be difficult to use Regex in OdfEdit search&replace functions, I will try to use it.

**Reply by eturpault** [reply]
I pushed a new version 2.5beta2 of OdfEdit.py in the branch v2.5_beta with the implementation of the improved search function (regular expression support, search range selection) and the addition of the replace feature. There is also the improvement of the PitchTuning calculation discussed with Oleg.
I can provide the Linux executable if someone wants it rather than the .py file.
